pkg/provisioning/import.go

```diff
@@ -0,0 +1,712 @@
+		}}
+	}
+
+	// we have to delete the old connector and create a new one
+	return []action{
+		deleteConnectorAction{
+			cfg:              oldConfig,
+			pipelineID:       pipelineID,
+			connectorService: s.connectorService,
+			pluginService:    s.pluginService,
+		},
+		createConnectorAction{
+			cfg:              newConfig,
+			pipelineID:       pipelineID,
+			connectorService: s.connectorService,
+			pluginService:    s.pluginService,
+		},
+	}
+}
+
+func (s *Service) prepareProcessorActions(oldConfig, newConfig config.Processor, parent processor.Parent) []action {
+	if oldConfig.ID == "" {
+		// no old config, it's a brand new processor
+		return []action{createProcessorAction{
+			cfg:              newConfig,
+			parent:           parent,
+			processorService: s.processorService,
+		}}
+	} else if newConfig.ID == "" {
+		// no new config, it's an old processor that needs to be deleted
+		return []action{deleteProcessorAction{
+			cfg:              oldConfig,
+			parent:           parent,
+			processorService: s.processorService,
+		}}
+	}
+
+	// first compare whole configs
+	if cmp.Equal(oldConfig, newConfig) {
+		// configs match, no need to do anything
+		return nil
+	}
+
+	// compare them again but ignore fields that can be updated, if configs
+	// are still different an update is not possible, we have to entirely
+	// recreate the processor
+	opts := []cmp.Option{
+		cmpopts.IgnoreFields(config.Processor{}, "Settings", "Workers"),
+	}
+	if cmp.Equal(oldConfig, newConfig, opts...) {
+		// only updatable fields don't match, we can update the processor
+		return []action{updateProcessorAction{
+			oldConfig:        oldConfig,
+			newConfig:        newConfig,
+			processorService: s.processorService,
+		}}
+	}
+
+	// we have to delete the old processor and create a new one
+	return []action{
+		deleteProcessorAction{
+			cfg:              oldConfig,
+			parent:           parent,
+			processorService: s.processorService,
+		},
+		createProcessorAction{
+			cfg:              newConfig,
+			parent:           parent,
+			processorService: s.processorService,
+		},
+	}
+}
+
+func (*Service) reverseActions(actions []action) {
+	for i, j := 0, len(actions)-1; i < j; i, j = i+1, j-1 {
+		actions[i], actions[j] = actions[j], actions[i]
+	}
+}
+
+type action interface {
+	String() string
+	Do(context.Context) error
+	Rollback(context.Context) error
+}
+
+// --------------------
+// -- CREATE ACTIONS --
+// --------------------
+
+type createPipelineAction struct {
+	cfg config.Pipeline
+
+	pipelineService PipelineService
+}
+
+func (a createPipelineAction) String() string {
+	return fmt.Sprintf("create pipeline with ID %v", a.cfg.ID)
+}
+func (a createPipelineAction) Do(ctx context.Context) error {
+	_, err := a.pipelineService.Create(ctx, a.cfg.ID, pipeline.Config{
+		Name:        a.cfg.Name,
+		Description: a.cfg.Description,
+	}, pipeline.ProvisionTypeConfig)
+	if err != nil {
+		return cerrors.Errorf("failed to create pipeline: %w", err)
```

Comment:
It would be nice if we added the pipeline ID to the error.

Reply:
Actually now I remembered I did not include the ID on purpose because this error will be wrapped by [`executeActions`](https://github.com/ConduitIO/conduit/blob/3e44152a877971cfb123b7f56a74180d34e2731b/pkg/provisioning/import.go#L162), the wrapped error will include the name of the action which [includes the ID](https://github.com/ConduitIO/conduit/blob/3e44152a877971cfb123b7f56a74180d34e2731b/pkg/provisioning/import.go#L353) as well as what the action is trying to do (create, update, delete).